### PR TITLE
New Sampler: DistributedWeightedRandomSampler

### PR DESCRIPTION
**Pull request by dani-capellan**
### Summary
This PR introduces `DistributedWeightedRandomSampler`, which combines the functionality of `WeightedRandomSampler `and `DistributedSampler`. This sampler enables weighted sampling across multiple distributed processes, ensuring each process receives a balanced subset of the weighted samples.

### Motivation
Currently, PyTorch does not provide a straightforward workaround to use `WeightedRandomSampler `in a distributed setting. This has been a common issue, see discussions below:

- [PyTorch Forums: WeightedRandomSampler + DistributedSampler](https://discuss.pytorch.org/t/weightedrandomsampler-distributedsampler/52817)
- [PyTorch Lightning Issue #10946](https://github.com/Lightning-AI/pytorch-lightning/issues/10946)

### Expected Usage

```
import torch
from torch.utils.data import DataLoader
from torch.utils.data.distributed import DistributedWeightedRandomSampler

dataset = list(range(10))
weights = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0]

sampler = DistributedWeightedRandomSampler(weights, num_samples=5, num_replicas=2, rank=0)
loader = DataLoader(dataset, sampler=sampler, batch_size=2)

for epoch in range(start_epoch, n_epochs):
    sampler.set_epoch(epoch)
    for batch in loader:
        print(batch)  # Each process gets a subset of the sampled indices
 ```
        
       

cc @andrewkho @divyanshk @SsnL @VitalyFedyunin @dzhulgakov @scotts

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -3315,6 +3315,8 @@
     "ConcatDataset",
     # torch.utils.data.distributed
     "DistributedSampler",
+    # torch.utils.data.distributed
+    "DistributedWeightedRandomSampler",
```
Comment (albanD) [inline docs/source/conf.py:3319]: Adding new entries in this file is not ok.
Please make sure this is properly documented on the website!

**Review by albanD — CHANGES_REQUESTED** [review]

**Comment by divyanshk** [comment]
@dani-capellan Can you add unit tests to `test/test_dataloader.py` ? Thanks.

**Comment by dani-capellan** [comment]
Hi. i already added some unit tests as requested. Thanks.

**Comment by bhack** [comment]
@dani-capellan Do you think any extension of this could let to have an online hard case sampler? See 
https://github.com/pytorch/data/issues/1493

I think it is mainly how to online collect and manipulate the weights in the training process.

**Comment by dani-capellan** [comment]
> @dani-capellan Do you think any extension of this could let to have an online hard case sampler? See [pytorch/data#1493](https://github.com/pytorch/data/issues/1493)
> 
> I think it is mainly how to online collect and manipulate the weights in the training process.

@bhack Yes, I'd say this class can be extended to implement online hard example sampling, maybe adding a method like `update_weights(self, new_weights: Sequence[float])` inside the class, which would be called during the training loop. The weights would evolve during training based on runtime difficulty estimates. However, i would maybe create another class for this, maybe inheriting from `DistributedWeightedRandomSampler`.

**Comment by bhack** [comment]
It would be quite useful.

File: torch/utils/data/distributed.py [hunk 2]
```diff
@@ -148,3 +148,117 @@ def set_epoch(self, epoch: int) -> None:
...
+        # Perform weighted sampling
+        sampled_indices = torch.multinomial(
+            weights, self.num_samples, self.replacement, generator=g
+        )
+        
+        # Map sampled indices back to original dataset indices
+        sampled_indices = indices[sampled_indices]
+
+        # Distribute samples across processes
+        # sampled_indices = sampled_indices[self.rank:self.num_samples_per_proc * self.num_replicas:self.num_replicas]
```
Comment (divyanshk) [inline torch/utils/data/distributed.py]: Extra commented line here.

File: torch/utils/data/distributed.py [hunk 3]
```diff
@@ -148,3 +148,117 @@ def set_epoch(self, epoch: int) -> None:
...
+    This sampler combines the behavior of `WeightedRandomSampler` (sampling based on weights)
+    and `DistributedSampler` (splitting data across multiple processes).
+
+    Args:
+        weights (sequence): A sequence of weights, not necessarily summing up to one.
+        num_samples (int): Total number of samples across all processes.
+        num_replicas (int, optional): Number of processes in distributed training.
+        rank (int, optional): Rank of the current process within num_replicas.
+        replacement (bool, optional): If True, samples are drawn with replacement.
+        shuffle (bool, optional): If True, shuffle indices before applying weights.
```
Comment (divyanshk) [inline torch/utils/data/distributed.py:165]: Not sure if I am following. If we shuffle then what is the point of the weights?

File: torch/utils/data/distributed.py [hunk 4]
```diff
@@ -148,3 +148,117 @@ def set_epoch(self, epoch: int) -> None:
...
+
+    This sampler combines the behavior of `WeightedRandomSampler` (sampling based on weights)
+    and `DistributedSampler` (splitting data across multiple processes).
+
+    Args:
+        weights (sequence): A sequence of weights, not necessarily summing up to one.
+        num_samples (int): Total number of samples across all processes.
+        num_replicas (int, optional): Number of processes in distributed training.
+        rank (int, optional): Rank of the current process within num_replicas.
+        replacement (bool, optional): If True, samples are drawn with replacement.
```
Comment (divyanshk) [inline torch/utils/data/distributed.py:164]: Would `replacement=True` mean we can get duplicate samples in two or more ranks ? That is would be quite different from how a typical `DistributedSampler` is expected to work.

File: torch/utils/data/distributed.py [hunk 5]
```diff
@@ -148,3 +148,117 @@ def set_epoch(self, epoch: int) -> None:
             epoch (int): Epoch number.
         """
         self.epoch = epoch
+
+
+class DistributedWeightedRandomSampler(Sampler[_T_co]):
+    """Sampler that applies weighted random sampling across multiple distributed processes.
```
Comment (divyanshk) [inline torch/utils/data/distributed.py:154]: Just so that I understand, would this mean that rank0 ends up getting all the samples with higher weights?

**Comment by divyanshk** [comment]
Left some comments to better understand. Also, can we explain the use-case of such a sampler a bit more ? Like in what situations would we need such a sampler - what is the current way to unblock that use-case, and the limitations of that approach? Thanks.

File: torch/utils/data/distributed.py (hunk 2 above)
Comment (dani-capellan) [reply torch/utils/data/distributed.py]: Removed in the new commit

File: torch/utils/data/distributed.py (hunk 3 above)
Comment (dani-capellan) [reply torch/utils/data/distributed.py:165]: The weights determine the sampling distribution in the multinomial step (torch.multinomial). If all weights are equal, sampling is uniform. The shuffle option simply permutes indices and their corresponding weights beforehand, to avoid ordering effects (e.g., if the dataset is sorted, high-weight samples might cluster and bias splits across ranks). Important: when shuffling, the weights are permuted consistently with the indices, so weights still control the distribution while shuffle only randomizes their order.

File: torch/utils/data/distributed.py (hunk 4 above)
Comment (dani-capellan) [reply torch/utils/data/distributed.py:164]: Yes, but with some caveats. With `replacement=True`, `torch.multinomial` can sample the same index multiple times, so duplicates may appear in the pool. Since this pool is then split across ranks, the same dataset index can end up on different ranks or appear multiple times within the same rank. This differs from `DistributedSampler`, which guarantees non-overlapping partitions. However, this is expected behavior for weighted sampling, as happens in `WeightedRandomSampler` in a non-distributed setting (there we have also this `replacement` option). You can consider this class as an intersection of `WeightedRandomSampler` and `DistributedSampler`, where these options can be adjusted as needed. In practice, this can be useful in some situations (and we have empirically observed benefits) when training on highly imbalanced datasets (e.g., in medical imaging, often combined with data augmentation). Nonetheless, as happens in `WeightedRandomSampler`, if this behavior is not desired, setting `replacement=False` gives behavior closer to `DistributedSampler`.

File: torch/utils/data/distributed.py (hunk 5 above)
Comment (dani-capellan) [reply torch/utils/data/distributed.py:154]: Not necessarily. Sampling is done globally across all samples using `torch.multinomial`. The weights guide which indices are drawn, but not which rank they go to. Then, after sampling, you split the sampled indices across ranks (start_index:end_index). This splitting is index-based, not weight-based. So, rank0 doesn’t inherently get the samples with higher weights. Each rank just gets its slice of the sampled pool. Whether rank0 sees more high-weight samples depends entirely on the ordering of the sampled pool after shuffling and multinomial draws, not in its rank.

**Comment by dani-capellan** [comment]
Hi @divyanshk,

Sorry for the delay in answering. I reviewed the code and answered to your comments.

I answer below to your question about the use-case of this sampler which, in my opinion, could be useful:

This sampler was originally motivated by work on lung nodule malignancy classification in medical imaging, where malignant (positive) cases are relatively rare (~1:10 compared to benign). In such highly imbalanced settings, using the standard `DistributedSampler` can lead to batches dominated by negatives, making training unstable and slow to converge. The usual workaround is `WeightedRandomSampler`, but it doesn’t support distributed training out of the box, so in multi-GPU runs, each process may oversample inconsistently or duplicate work.

This class combines both ideas: weighted sampling (to upsample positives or any minority class) and distributed splitting (to ensure each rank gets its fair share of the weighted pool). This avoids skew across GPUs while still tackling imbalance.

Limitations: it can introduce duplicates if `replacement=True`, and it changes the strict partitioning guarantees of `DistributedSampler`. But, e.g., for imbalanced medical datasets, it’s often a practical trade-off.

Please let me know if you need further clarifications, and once again, sorry for the delay.

Best,

Daniel
